### PR TITLE
Implemented nesting support in document outline.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
Before nested headings were totally ignored now we show them with a representation of their nesting path.
This PR also updates TableOfContentsPanel counting methodology to count nested blocks.
Two small functions computeOutlineHeadings, and blockCount were implemented, should we expose them as an util? I decided not to do so because I don't see other usages for them and if the usages appear later we can always extract them.

## How Has This Been Tested?
Verify documented outline and the countings panel work as expected, e.g: the warnings are still displayed and the countings are correct.
Try to nest headings inside columns and verify they appear in the document outline, with their nesting path displayed.

## Screenshots (jpeg or gifs if applicable):

image

### CODE REVIEW

File: editor/components/document-outline/index.js
Comment (mcsf) [inline editor/components/document-outline/index.js]: Better with `lodash#isEmpty`

File: editor/components/document-outline/test/index.js
Comment (mcsf) [inline editor/components/document-outline/test/index.js]: Maybe a slightly more complex test would be a good complement.

**Review by mcsf — COMMENTED** [review]
This implementation looks like a good start. I'm wondering, though, if there is some parent issue or any discussion somewhere on this topic.

A couple of thoughts on my mind:
- Will there be justifiable scenarios where heading levels have different weights by virtue of belonging to a nested context?
- What does nesting (and notably multidimensional nesting) mean for the flow of the document, and thus its outline?

File: editor/components/document-outline/index.js
Comment (jorgefilipecosta) [reply editor/components/document-outline/index.js]: Improved.

File: editor/components/document-outline/test/index.js
Comment (jorgefilipecosta) [reply editor/components/document-outline/test/index.js]: The test case was extended 👍

**Comment by jorgefilipecosta** [comment]
Hi @mcsf,

> A couple of thoughts on my mind:

> Will there be justifiable scenarios where heading levels have different weights by virtue of belonging to a nested context?
>What does nesting (and notably multidimensional nesting) mean for the flow of the document, and thus its outline?

These are really interesting questions. W3.org has a section about outlines and they even provide an algorithm to correctly implement them https://www.w3.org/TR/2011/WD-html5-20110525/sections.html#outlines.

As a very short summary, headings inside these elements "blockquote body details fieldset figure td" (section root) should not count or appear in the parent document outline.
Nesting inside these elements "article aside nav section" (section content) scopes the headings e.g: they should start again from level 1 but they are lower one level the level of the parent of this element.

Provided right now we can not nest headings inside section root or section content, from the outline algorithm point of view there is no nesting. So it looks like right now nesting does not change weights. We are just showing the nesting path in the outline as a way to make the structure more intuitive to the user, and make it easier to navigate in the document.

As nesting involves we should keep eye on this.

File: editor/components/document-outline/index.js
Comment (aduth) [inline editor/components/document-outline/index.js]: Is the `isEmpty` check necessary? `flatMap` on an empty array should return an empty array anyways.

File: editor/components/document-outline/item.js
Comment (aduth) [inline editor/components/document-outline/item.js]: Tab before parenthesis should be a space.

File: editor/components/table-of-contents/panel.js
Comment (aduth) [inline editor/components/table-of-contents/panel.js]: Should this be a propers elector?

File: editor/components/table-of-contents/panel.js
Comment (aduth) [inline editor/components/table-of-contents/panel.js]: This logic is excessively complex to be inlined without comments or some named variable / separate composable function.

File: editor/components/table-of-contents/panel.js
Comment (jorgefilipecosta) [reply editor/components/table-of-contents/panel.js]: Yes, it is a good idea and has some advantages. Namely, we can make use of the state.editor.present.blocksByUid structure which is flat so we can have a much simpler and efficient algorithm when compared with the previous recursive version.

File: editor/components/document-outline/index.js
Comment (gziolo) [inline editor/components/document-outline/index.js]: As discussed on Slack today. Can we convert this functionality (`computeOutlineHeadings `) into a general purpose selector `getBlocksOfType`?

Comment (jorgefilipecosta) [reply editor/components/document-outline/index.js]: Unfortunately, at the time of our discussion, I did not remember that here we don't want just the heading blocks we also want the path to it in order to display it to the user.
The path is the array of ancestors of a block until a heading. so it can be ["Core/collumn"] or [] or ["cover/image", "Core/collumn"].
I think this is the only place where we need to retrieve the blocks of a given type sorted by document position with its ancestors so this function is probably not reusable elsewhere.

File: editor/components/document-outline/item.js
Comment (jorgefilipecosta) [reply editor/components/document-outline/item.js]: Hi @gziolo, I just described what path is in comment https://github.com/WordPress/gutenberg/pull/5314#discussion_r177419949.

File: editor/store/selectors.js
Comment (jorgefilipecosta) [reply editor/store/selectors.js]: Thank you for catching this it was fixed.

File: editor/components/document-outline/item.js
Comment (gziolo) [reply editor/components/document-outline/item.js]: Yes, but we need to have it better explained in the code for the future us editing this code :)

Comment (jorgefilipecosta) [reply editor/components/document-outline/item.js]: Yes, you are right I added comments on this line and for the function computeOutlineHeadings where the path is computed :)

**Comment by jorgefilipecosta** [comment]
This PR was rebased and should be ready to merge soon If we don't find any blocking point.

**Comment by jorgefilipecosta** [comment]
Hi @karmatosed could you please do a final verification regarding the UX/design before the merge :)?

**Comment by karmatosed** [comment]
I think for now this is good but also agree with this:

> As nesting involves we should keep eye on this.

We should really see and be ok revisiting this UI also.

**Review by karmatosed — APPROVED** [review]

File: editor/components/document-outline/index.js
Comment (aduth) [inline editor/components/document-outline/index.js]: JSDoc type should be `?Array`

File: editor/components/document-outline/item.js
Comment (aduth) [inline editor/components/document-outline/item.js]: Minor: There is a `BlockTitle` component now, which could either be accessed by its connected form (pass UID), or its raw form taking `name` as a prop.

https://github.com/WordPress/gutenberg/tree/master/editor/components/block-title

**Review by aduth — APPROVED** [review]

**Comment by paulwilde** [comment]
Yeah seems the UI would need to be redesigned if the nesting goes beyond 2 levels deep. Here is an example of a website (with the text changed to Lorem ipsum) which uses columns nested into sections:

screen shot 2018-04-06 at 18 49 11

**Comment by afercia** [comment]
Looking at this one year later :) Worth noting the [W3C outline algorithm](https://www.w3.org/TR/2011/WD-html5-20110525/sections.html#outlines) mentioned in https://github.com/WordPress/gutenberg/pull/5314#issuecomment-370598719  cannot be relied upon to build an outline. The linked resource is outdated and more importantly there are no known implementations of the algorithm.

Up-to-date W3C recommendation in HTML 5.2:
https://www.w3.org/TR/html5/sections.html#creating-an-outline
> Creating an outline
> Warning! There are currently no known native implementations of the outline algorithm in graphical browsers or assistive technology user agents, although the algorithm is implemented in other software such as conformance checkers and browser extensions. Therefore **the [outline](https://www.w3.org/TR/html5/sections.html#outline) algorithm cannot be relied upon to convey document structure** to users. **Authors should use heading rank (h1-h6) to convey document structure**.

Regardless, I'm not sure displaying the heading ancestors adds value for users. Personally, I find it confusing. Will create a new issue.

**Pull request merged**
